csa.swift

```diff
@@ -0,0 +1,125 @@
+    }
+}
+
+struct Router {
+    private let connections: [Connection]
+    private let count: Int
+    
+    init(input: InputType) throws {
+        var connections: [Connection] = []
+        while let connection = try Connection(input: input) {
+            connections.append(connection)
+        }
+        self.connections = connections
+        self.count = connections.reduce(0) { (count, connection) in
+            max(count, connection.from + 1, connection.to + 1)
+        }
+    }
+    
+    func solve(request: Request) -> [Connection] {
+        let from = request.from, to = request.to, departure = request.departure
+        guard count > max(from, to) else { return [] }
+        var incoming: [Connection?] = Array(count: count, repeatedValue: nil)
+        for connection in connections {
+            if connection.departure < departure || connection.to == from {
+                continue
+            }
+            if let final = incoming[to] where connection.departure > final.arrival {
+                break
+            }
+            if connection.from != from {
+                guard let previous = incoming[connection.from] where connection.departure > previous.arrival else {
+                    continue
+                }
+            }
+            if let current = incoming[connection.to] where connection.arrival > current.arrival {
+                continue
+            }
+            incoming[connection.to] = connection
+        }
+        var route: [Connection] = []
+        var station = to
+        while let connection = incoming[station] {
+            route.insert(connection, atIndex: 0)
+            station = connection.from
+        }
+        return route
+    }
+}
+
+struct TextualInput: InputType {
+    private let body: () -> String?
+    
+    init(_ body: () -> String?) {
+        self.body = body
+    }
+    
+    func read(count: Int) throws -> [Int]? {
+        guard let line = body() else { return nil }
+        guard line.characters.count > 0 else { return nil }
+        let components = line.componentsSeparatedByString(" ")
+        guard components.count == count else { throw Error("There should be exactly \(count) values on this line.") }
+        return try components.map { component in
+            guard let integer = Int(component) else { throw Error("\"\(component)\" is not an integer.") }
+            return integer
+        }
+    }
+}
+
+func printRoute<S: SequenceType where S.Generator.Element == Connection>(route: S) {
+    for connection in route {
+        print("\(connection.from) \(connection.to) \(connection.departure) \(connection.arrival)")
+    }
+    print("")
+    fflush(stdout)
+}
+
+do {
+    let input = TextualInput({ readLine() })
+    let router = try Router(input: input)
+    while let request = try Request(input: input) {
+        let route = router.solve(request)
+        printRoute(route)
+    }
+} catch let error as Error {
+    fputs("\(error.message)\n", stdout)
```

Comment:
It’s not really part of the spec, but `stderr` would probably be more appropriate here.